test/unit/test_environment.py

```diff
@@ -0,0 +1,160 @@
+                 'sagemaker_submit_directory': 'imagenet', 'sagemaker_enable_cloudwatch_metrics': True,
+                 'sagemaker_container_log_level': logging.WARNING}
+
+ALL_HPS = dict(chain(USER_HPS.items(), SAGEMAKER_HPS.items()))
+
+
+@pytest.fixture(name='opt_ml_path')
+def override_opt_ml_path(tmpdir):
+    opt_ml = tmpdir.mkdir('opt').mkdir('ml')
+    with patch.dict('os.environ', {'BASE_PATH': str(opt_ml)}):
+        reload_module(environment)
+        yield opt_ml
+    reload_module(environment)
+
+
+@pytest.fixture(name='input_path')
+def override_input_path(opt_ml_path):
+    return opt_ml_path.mkdir('input')
+
+
+@pytest.fixture(name='input_config_path')
+def override_input_config_path(input_path):
+    return input_path.mkdir('config')
+
+
+@pytest.fixture(name='input_data_path')
+def override_input_data_path(input_path):
+    return input_path.mkdir('data')
+
+
+def test_read_json(tmpdir):
+    file_path = write_json('hyperparameters.json', ALL_HPS, tmpdir)
+
+    assert environment.read_json(file_path) == ALL_HPS
+
+
+def test_read_json_throws_exception():
+    with pytest.raises(IOError):
+        environment.read_json('non-existent.json')
+
+
+def test_read_hyperparameters(input_config_path):
+    write_json('hyperparameters.json', ALL_HPS, input_config_path)
+
+    assert environment.read_hyperparameters() == ALL_HPS
+
+
+def test_read_key_serialized_hyperparameters(input_config_path):
+    key_serialized_hps = {k: json.dumps(v) for k, v in ALL_HPS.items()}
+    write_json('hyperparameters.json', key_serialized_hps, input_config_path)
+
+    assert environment.read_hyperparameters() == ALL_HPS
+
+
+def test_split_hyperparameters_only_provided_by_user():
+    assert environment.split_hyperparameters(USER_HPS) == ({}, USER_HPS)
+
+
+def test_split_hyperparameters_only_provided_by_sagemaker():
+    assert environment.split_hyperparameters(SAGEMAKER_HPS) == (SAGEMAKER_HPS, {})
+
+
+def test_split_hyperparameters():
+    assert environment.split_hyperparameters(ALL_HPS) == (SAGEMAKER_HPS, USER_HPS)
+
+
+def test_resource_config(input_config_path):
+    write_json('resourceconfig.json', RESOURCE_CONFIG, input_config_path)
+
+    assert environment.read_resource_config() == RESOURCE_CONFIG
+
+
+def test_input_data_config(input_config_path):
+    write_json('inputdataconfig.json', INPUT_DATA_CONFIG, input_config_path)
+
+    assert environment.read_input_data_config() == INPUT_DATA_CONFIG
+
+
+def test_channel_input_dirs(input_data_path):
+    assert environment.channel_path('evaluation') == str(input_data_path.join('evaluation'))
+    assert environment.channel_path('training') == str(input_data_path.join('training'))
+
+
+@patch('subprocess.check_output', lambda s: b('GPU 0\nGPU 1'))
+def test_gpu_count_in_gpu_instance():
+    assert environment.gpu_count() == 2
+
+
+def test_gpu_count_in_cpu_instance():
+    assert environment.gpu_count() == 0
```

Comment:
if I run this in a box with an nvidia gpu + nvidia-smi installed it will fail. This is not that uncommon.

Reply:
Good point, patching it as well.